File: CS50/credit/credit.c
// Importing all the required libraries
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef char *string; // Creating a string variable
string input_function();
int summer(int num);

int main(void) // Main method
{
  int a[20]; // An array of 20 integers
  int sum = 0;
  string s = input_function(); // Storing the input in this variable
  int len = strlen(s);
  for (int i = 0; i < len - 1;
       i += 2) // An algorithm to multiply the appropriate digits by 2 and
               // convert into integer array
  {
    if (len % 2 != 0) // if length is odd Eg.5
    {
      a[i] = (s[i] - '0');
      a[i + 1] = ((s[i + 1] - '0') * 2);
    } else // even
    {
      a[i] = ((s[i] - '0') * 2);
      a[i + 1] = ((s[i + 1] - '0'));
    }
  }
  if (len % 2 != 0) // adding the last digit to the array to not make it skip
  {
    a[len - 1] = (s[len - 1] - '0');
  }
  for (int j = 0; j < len; j++) // summing two digit number in the array
  {
    a[j] = summer(a[j]);
  }
  for (int j = 0; j < len; j++) // summing all the numbers in the array
  {
    sum += a[j];
  }
  if ((sum % 10 == 0) && (s[0] == '3') && ((s[1] == '4') || (s[1] == '7')) &&
      (len == 15)) // Checking for American Express
  {
    printf("AMEX\n");
  } else if ((len == 16) && (s[0] == '5') && (sum % 10 == 0) &&
             ((s[1] == '1') || (s[1] == '2') || (s[1] == '3') ||
              (s[1] == '4') || (s[1] == '5'))) // Checking for mastercard
  {
    printf("MASTERCARD\n");
  } else if (((len == 13) || (len == 16)) && (sum % 10 == 0) &&
             (s[0] == '4')) // Checking for visa

  {
    printf("VISA\n");
  } else // the invalid output option
  {
    printf("INVALID\n");
  }
  free(s); // freeing the allocated memory by malloc
}

string input_function() // Failproof methond to take input from the user
{
  string si =
      malloc(sizeof(char) * 20); // Allocation of the char array of 20 bytes
  int i;
  long val;
  char *next;
  int h = 1;
  while (h) // Can optimize this part by using ASCII values for conversion
  {
    printf("Number: ");
    scanf("%s", si);
    val = strtol(si, &next, 10);
    if ((next == si) || (*next != '\0')) {
      h = 1;
    } else {
      h = 0;
    }
  }
  return si;
}

int summer(int num) // A function to sum all the digits in a number and return
                    // the value
{
  int sum = 0;
  while (num > 0) {
    sum += num % 10;
    num /= 10;
  }
  return sum;
}

File: CS50/sentimental-credit/credit.py
# TODO
# A function to find the checksum value of a list
def checksum(list):
    copy = []
    for x in range(len(list)):  # copying and creating a new list
        copy.append(list[x])
    flag = False
    sum = 0
    for x in range(2, len(copy)+1, 2):  # multiplytin by 2
        copy[-x] = copy[-x] * 2

    for x in range(0, len(list), 1):
        copy[x] = summer(copy[x])

    for x in range(len(copy)):  # finding sum
        sum += copy[x]

    if sum % 10 == 0:  # checksum
        flag = True
    return flag

# A function to sum two digit number in array


def summer(number):
    num = str(number)
    sum = 0
    for x in range(len(num)):
        sum = sum + int(num[x])
    return sum


def inp():  # A fucntion to take proper input from the user
    list = []
    inp = None
    while True:
        inp = input("Number: ")
        if inp.isnumeric():
            break
    for x in inp:
        list.append(int(x))
    return list


list = inp()
flag = checksum(list)
leng = len(list)

if flag and leng == 15 and list[0] == 3 and (list[1] == 4 or list[1] == 7):  # checking for amex
    print("AMEX")

elif flag and leng == 16 and list[0] == 5 and (list[1] > 0 and list[1] < 6):  # chekcing for mastercard
    print("MASTERCARD")

elif flag and (leng == 13 or leng == 16) and list[0] == 4:  # cheking for visa
    print("VISA")
else:  # else case invalid
    print("INVALID")

